src/NumericalAlgorithms/Spectral/Chebyshev.cpp

```diff
@@ -0,0 +1,121 @@
+  // Algorithm 21 from Kopriva's book, p. 60
+  // Since the trigonometric form is expensive to compute we make use of it only
+  // for large k. It remains to be determined what is a reasonable crossover
+  // point between the two forms, e.g. depending on compiler settings.
+  static constexpr size_t chebyshev_recursion_to_trigonometry_crossover = 10;
+  switch (k) {
+    case 0:
+      return DataVector(x.size(), 1.);
+    case 1:
+      return x;
+    default:
+      if (k <= chebyshev_recursion_to_trigonometry_crossover) {
+        DataVector T_k_minus_2(x.size(), 1.);
+        DataVector T_k_minus_1 = x;
+        DataVector T_k(x.size());
+        for (size_t j = 2; j <= k; j++) {
+          T_k = 2 * x * T_k_minus_1 - T_k_minus_2;
+          T_k_minus_2 = T_k_minus_1;
+          T_k_minus_1 = T_k;
+        }
+        return T_k;
+      } else {
+        return cos(k * acos(x));
+      }
+  }
+}
+
+template <Basis>
+DataVector compute_inverse_weight_function_values(const DataVector&) noexcept;
+template <>
+DataVector compute_inverse_weight_function_values<Basis::Chebyshev>(
+    const DataVector& x) noexcept {
+  return sqrt(1. - square(x));
+}
+
+template <Basis>
+double compute_basis_function_normalization_square(size_t) noexcept;
+template <>
+double compute_basis_function_normalization_square<Basis::Chebyshev>(
+    const size_t k) noexcept {
+  if (k == 0) {
+    return M_PI;
+  } else {
+    return M_PI_2;
+  }
+}
+
+template <Basis, Quadrature>
+std::pair<DataVector, DataVector> compute_collocation_points_and_weights(
+    size_t) noexcept;
+/// \endcond
+
+// Algorithm to compute Chebyshev-Gauss quadrature
+
+/// \cond
+template <>
+std::pair<DataVector, DataVector>
+compute_collocation_points_and_weights<Basis::Chebyshev, Quadrature::Gauss>(
+    const size_t num_points) noexcept {
+  // Algorithm 26 from Kopriva's book, p. 67
+  ASSERT(num_points >= 1,
+         "Chebyshev-Gauss quadrature requires at least one collocation point.");
+  const size_t poly_degree = num_points - 1;
+  DataVector x(num_points);
+  const DataVector w(num_points, M_PI / num_points);
+  for (size_t j = 0; j < num_points; j++) {
+    x[j] = -cos((2. * double(j) + 1.) / (2. * poly_degree + 2.) * M_PI);
+  }
+  return std::make_pair(std::move(x), w);
+}
+/// \endcond
+
+// Algorithm to compute Chebyshev-Gauss-Lobatto quadrature
+
+/// \cond
+template <>
+std::pair<DataVector, DataVector> compute_collocation_points_and_weights<
+    Basis::Chebyshev, Quadrature::GaussLobatto>(
+    const size_t num_points) noexcept {
+  // Algorithm 27 from Kopriva's book, p. 68
+  ASSERT(num_points >= 2,
+         "Chebyshev-Gauss-Lobatto quadrature requires at least two collocation "
+         "points.");
+  const size_t poly_degree = num_points - 1;
+  DataVector x(num_points);
+  DataVector w(num_points, M_PI / poly_degree);
+  for (size_t j = 0; j < num_points; j++) {
+    x[j] = -cos(double(j) / double(poly_degree) * M_PI);
+  }
+  w[0] /= 2.;
```

Comment:
ok but why :D

Reply:
multiplication is much faster than division

Reply:
is this not something the compiler can cope with? anyway, changed it

Reply:
It can in theory handle it as long as the LHS is a fundamental and the RHS is a literal power of two.  (I expect compilers actually do do that transformation but I haven't checked.)  These conditions are met in this case, but it is easier to remember and check a rule of "never divide by a literal double" than the rule with the two part exception above, and as computational physicists we are used to interpreting 0.5, 0.25, and similar simple fractions.